server/src/test/java/org/elasticsearch/cluster/routing/allocation/allocator/ClusterAllocationSimulationTests.java

```diff
@@ -0,0 +1,522 @@
+
+@TestLogging(reason = "progress reporting", value = "org.elasticsearch.cluster.routing.allocation.allocator.DesiredBalanceComputer:DEBUG")
+public class ClusterAllocationSimulationTests extends ESAllocationTestCase {
+
+    private static final Logger logger = LogManager.getLogger(ClusterAllocationSimulationTests.class);
+
+    private static int nodeCount(long tierBytes, long nodeBytes, double tierWriteLoad, double nodeIndexingThreads, int minCount) {
+        var watermarkNodeBytes = (long) (nodeBytes * 0.85);
+        return Math.max(
+            Math.toIntExact((tierBytes + watermarkNodeBytes - 1) / watermarkNodeBytes),
+            Math.max((int) Math.ceil(tierWriteLoad / nodeIndexingThreads), minCount)
+        ) + between(0, 2);
+    }
+
+    public void testBalanceQuality() {
+
+        final var shardSizesByIndex = new HashMap<String, Long>();
+        final var tiers = new String[] { DataTier.DATA_HOT, DataTier.DATA_WARM, DataTier.DATA_COLD };
+        final var tierSizes = Arrays.stream(tiers).collect(Collectors.toMap(Function.identity(), s -> 0L));
+        final var tierWriteLoads = Arrays.stream(tiers).collect(Collectors.toMap(Function.identity(), s -> 0.0));
+        final var maxSizeVariance = ByteSizeValue.ofMb(20).getBytes();
+
+        final var metadataBuilder = Metadata.builder();
+        for (var dataStreamIndex = between(10, 50); dataStreamIndex >= 0; dataStreamIndex -= 1) {
+            final var hotIndices = between(2, 20);
+            final var warmIndices = between(10, 30);
+            final var coldIndices = between(10, 100);
+            final var hotShards = randomFrom(1, 2, 4, 8);
+            final var shrunkShards = randomFrom(1, 2);
+            final var approxIndexSize = randomFrom(ByteSizeValue.ofGb(1), ByteSizeValue.ofGb(10), ByteSizeValue.ofGb(50));
+
+            final var indexCount = hotIndices + warmIndices + coldIndices;
+            for (var index = 0; index < indexCount; index++) {
+                final var tier = index < coldIndices ? DataTier.DATA_COLD
+                    : index < coldIndices + warmIndices ? DataTier.DATA_WARM
+                    : DataTier.DATA_HOT;
+                final var indexName = Strings.format("index-%03d-%s-%03d", dataStreamIndex, tier.charAt(5), index);
+                final var shardCount = DataTier.DATA_HOT.equals(tier) ? hotShards : shrunkShards;
+                final var replicaCount = DataTier.DATA_COLD.equals(tier) ? 0 : 1;
+
+                final var indexWriteLoadBuilder = IndexWriteLoad.builder(shardCount);
+                var totalWriteLoad = 0.0;
+                for (int shardId = 0; shardId < shardCount; shardId++) {
+                    final var shardWriteLoad = index == indexCount - 1 ? (scaledRandomIntBetween(1, 8000) / 1000.0) : 0.0;
+                    totalWriteLoad += shardWriteLoad;
+                    indexWriteLoadBuilder.withShardWriteLoad(shardId, shardWriteLoad, TimeValue.timeValueHours(1).millis());
+                }
```

Comment:
Once we have the proper pre-calculated write-load meta-data in place that @fcofdez  is working on we should simply set that instead - and set it to 0 for the non-write indices.